### PR TITLE
Ble-4: Split Launch Fragment

### PR DESCRIPTION
**Pull request by SamiraVaezi**
## Description
The operation of checking permissions and other preconditions have been transferred to the repository and a ViewModel is created to emit the result into livedata. Now the fragment can relax and just listen to updates.

### CODE REVIEW

File: app/src/main/java/net/grandcentrix/blutufintegration/data/repo/BluetoothRepository.kt [hunk 1]
```diff
@@ -100,10 +99,8 @@ class BluetoothRepository {
         }
     }
 
-    fun deviceSupportsBluetooth(context: Context): Boolean {
-        val bluetoothManager =
-            context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager?
-        return bluetoothManager?.adapter != null
+    fun checkPreconditions(): List<MissingPrecondition> {
```
Comment (jreehuis-gcx) [inline app/src/main/java/net/grandcentrix/blutufintegration/data/repo/BluetoothRepository.kt]: This change is a really good step.
Ideally you should have your own `enum` or `sealed` class for a missing precondition. Using the `Blutuf.MissingPrecondition` would expose `Blutuf` to the user of the `BluetoothRepository`. But the idea is that the `BluetoothRepository` hides `Blutuf` from its users.

File: app/src/main/java/net/grandcentrix/blutufintegration/ui/launch/LaunchFragment.kt [hunk 2]
```diff
@@ -1,36 +1,30 @@
...
-import org.koin.android.ext.android.inject
-
+import org.koin.androidx.viewmodel.ext.android.viewModel
 
 private const val REQUEST_CODE_PERMISSION_LOCATION = 1
 
-
 class LaunchFragment : Fragment() {
 
-    @VisibleForTesting
```
Comment (jreehuis-gcx) [inline app/src/main/java/net/grandcentrix/blutufintegration/ui/launch/LaunchFragment.kt:30]: When removing this annotation the property should be `private`.

File: app/src/main/java/net/grandcentrix/blutufintegration/ui/launch/LaunchFragment.kt [hunk 3]
```diff
@@ -46,53 +40,35 @@ class LaunchFragment : Fragment() {
...
-    fun checkPermissions(): CheckResult {
-        val result = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
-            permissions.any {
-                checkSelfPermission(requireContext(), it) !=
-                        PackageManager.PERMISSION_GRANTED
+    private fun updateUi(conditions: List<MissingPrecondition>) {
+        Log.e("sami", "error: ${conditions.size}")
+
+        binding.checkboxPermission.isChecked = true
```
Comment (jreehuis-gcx) [inline app/src/main/java/net/grandcentrix/blutufintegration/ui/launch/LaunchFragment.kt]: Setting the `Checkbox` state here without any condition but later setting the value with a condition is double work.
```suggestion
        binding.checkboxPermission.isChecked = conditions.all { cond -> cond !is FineLocationPermissionMissingError }
```

File: app/src/main/java/net/grandcentrix/blutufintegration/ui/launch/LaunchFragment.kt [hunk 4]
```diff
@@ -46,53 +40,35 @@ class LaunchFragment : Fragment() {
...
+        binding.checkboxBle.isChecked = true
```
Comment (jreehuis-gcx) [inline app/src/main/java/net/grandcentrix/blutufintegration/ui/launch/LaunchFragment.kt]: Setting the `Checkbox` state here without any condition but later setting the value with a condition is double work.
```suggestion
        binding.checkboxBle.isChecked = conditions.all { cond -> cond !is BluetoothDisabledError }
```

File: app/src/main/java/net/grandcentrix/blutufintegration/ui/launch/LaunchFragment.kt [hunk 5]
```diff
@@ -46,53 +40,35 @@ class LaunchFragment : Fragment() {
...
+        updateStartBtn(false)
```
Comment (jreehuis-gcx) [inline app/src/main/java/net/grandcentrix/blutufintegration/ui/launch/LaunchFragment.kt]: Setting the `Button` state here without any condition but later setting the value with a condition is double work.
```suggestion
        updateStartBtn(conditions.any { cond -> cond is FineLocationPermissionMissingError })
```

**Review by jreehuis-gcx — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/grandcentrix/blutufintegration/data/repo/BluetoothRepository.kt (hunk 1 above)
Comment (SamiraVaezi) [reply app/src/main/java/net/grandcentrix/blutufintegration/data/repo/BluetoothRepository.kt]: You’re right, I'll fix it.

**Review by jreehuis-gcx — APPROVED** [review]

**Pull request merged**